// @drift-labs/sdk/lib/constants/spotMarkets.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SpotMarkets = exports.MainnetSpotMarkets = exports.DevnetSpotMarkets = exports.WRAPPED_SOL_MINT = void 0;
const web3_js_1 = require("@solana/web3.js");
const __1 = require("../");
const numericConstants_1 = require("./numericConstants");
exports.WRAPPED_SOL_MINT = new web3_js_1.PublicKey('So11111111111111111111111111111111111111112');
exports.DevnetSpotMarkets = [
    {
        symbol: 'USDC',
        marketIndex: 0,
        oracle: new web3_js_1.PublicKey('5SSkXsEKQepHHAewytPVwdej4epN1nxgLVM84L4KXgy7'),
        oracleSource: __1.OracleSource.PYTH_STABLE_COIN,
        mint: new web3_js_1.PublicKey('8zGuJQqwhZafTah7Uc7Z4tXRnguqkn5KLFAP8oV6PHe2'),
        precision: new __1.BN(10).pow(numericConstants_1.SIX),
        precisionExp: numericConstants_1.SIX,
    },
    {
        symbol: 'SOL',
        marketIndex: 1,
        oracle: new web3_js_1.PublicKey('J83w4HKfqxwcq3BEMMkPFSppX3gqekLyLJBexebFVkix'),
        oracleSource: __1.OracleSource.PYTH,
        mint: new web3_js_1.PublicKey(exports.WRAPPED_SOL_MINT),
        precision: numericConstants_1.LAMPORTS_PRECISION,
        precisionExp: numericConstants_1.LAMPORTS_EXP,
        serumMarket: new web3_js_1.PublicKey('8N37SsnTu8RYxtjrV9SStjkkwVhmU8aCWhLvwduAPEKW'),
        phoenixMarket: new web3_js_1.PublicKey('78ehDnHgbkFxqXZwdFxa8HK7saX58GymeX2wNGdkqYLp'),
    },
    {
        symbol: 'BTC',
        marketIndex: 2,
        oracle: new web3_js_1.PublicKey('HovQMDrbAgAYPCmHVSrezcSmkMtXSSUsLDFANExrZh2J'),
        oracleSource: __1.OracleSource.PYTH,
        mint: new web3_js_1.PublicKey('3BZPwbcqB5kKScF3TEXxwNfx5ipV13kbRVDvfVp5c6fv'),
        precision: new __1.BN(10).pow(numericConstants_1.SIX),
        precisionExp: numericConstants_1.SIX,
        serumMarket: new web3_js_1.PublicKey('AGsmbVu3MS9u68GEYABWosQQCZwmLcBHu4pWEuBYH7Za'),
    },
];
exports.MainnetSpotMarkets = [
    {
        symbol: 'USDC',
        marketIndex: 0,
        oracle: new web3_js_1.PublicKey('Gnt27xtC473ZT2Mw5u8wZ68Z3gULkSTb5DuxJy7eJotD'),
        oracleSource: __1.OracleSource.PYTH_STABLE_COIN,
        mint: new web3_js_1.PublicKey('EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v'),
        precision: numericConstants_1.QUOTE_PRECISION,
        precisionExp: numericConstants_1.QUOTE_PRECISION_EXP,
    },
    {
        symbol: 'SOL',
        marketIndex: 1,
        oracle: new web3_js_1.PublicKey('H6ARHf6YXhGYeQfUzQNGk6rDNnLBQKrenN712K4AQJEG'),
        oracleSource: __1.OracleSource.PYTH,
        mint: new web3_js_1.PublicKey(exports.WRAPPED_SOL_MINT),
        precision: numericConstants_1.LAMPORTS_PRECISION,
        precisionExp: numericConstants_1.LAMPORTS_EXP,
        serumMarket: new web3_js_1.PublicKey('8BnEgHoWFysVcuFFX7QztDmzuH8r5ZFvyP3sYwn1XTh6'),
        phoenixMarket: new web3_js_1.PublicKey('4DoNfFBfF7UokCC2FQzriy7yHK6DY6NVdYpuekQ5pRgg'),
    },
    {
        symbol: 'mSOL',
        marketIndex: 2,
        oracle: new web3_js_1.PublicKey('E4v1BBgoso9s64TQvmyownAVJbhbEPGyzA3qn4n46qj9'),
        oracleSource: __1.OracleSource.PYTH,
        mint: new web3_js_1.PublicKey('mSoLzYCxHdYgdzU16g5QSh3i5K3z3KZK7ytfqcJm7So'),
        precision: new __1.BN(10).pow(numericConstants_1.NINE),
        precisionExp: numericConstants_1.NINE,
        serumMarket: new web3_js_1.PublicKey('9Lyhks5bQQxb9EyyX55NtgKQzpM4WK7JCmeaWuQ5MoXD'),
    },
    {
        symbol: 'wBTC',
        marketIndex: 3,
        oracle: new web3_js_1.PublicKey('GVXRSBjFk6e6J3NbVPXohDJetcTjaeeuykUpbQF8UoMU'),
        oracleSource: __1.OracleSource.PYTH,
        mint: new web3_js_1.PublicKey('3NZ9JMVBmGAqocybic2c7LQCJScmgsAZ6vQqTDzcqmJh'),
        precision: new __1.BN(10).pow(numericConstants_1.EIGHT),
        precisionExp: numericConstants_1.EIGHT,
        serumMarket: new web3_js_1.PublicKey('3BAKsQd3RuhZKES2DGysMhjBdwjZYKYmxRqnSMtZ4KSN'),
    },
    {
        symbol: 'wETH',
        marketIndex: 4,
        oracle: new web3_js_1.PublicKey('JBu1AL4obBcCMqKBBxhpWCNUt136ijcuMZLFvTP7iWdB'),
        oracleSource: __1.OracleSource.PYTH,
        mint: new web3_js_1.PublicKey('7vfCXTUXx5WJV5JADk17DUJ4ksgau7utNKj4b963voxs'),
        precision: new __1.BN(10).pow(numericConstants_1.EIGHT),
        precisionExp: numericConstants_1.EIGHT,
        serumMarket: new web3_js_1.PublicKey('BbJgE7HZMaDp5NTYvRh5jZSkQPVDTU8ubPFtpogUkEj4'),
        phoenixMarket: new web3_js_1.PublicKey('Ew3vFDdtdGrknJAVVfraxCA37uNJtimXYPY4QjnfhFHH'),
    },
    {
        symbol: 'USDT',
        marketIndex: 5,
        oracle: new web3_js_1.PublicKey('3vxLXJqLqF3JG5TCbYycbKWRBbCJQLxQmBGCkyqEEefL'),
        oracleSource: __1.OracleSource.PYTH_STABLE_COIN,
        mint: new web3_js_1.PublicKey('Es9vMFrzaCERmJfrF4H2FYD4KCoNkY11McCe8BenwNYB'),
        precision: numericConstants_1.QUOTE_PRECISION,
        precisionExp: numericConstants_1.QUOTE_PRECISION_EXP,
        serumMarket: new web3_js_1.PublicKey('B2na8Awyd7cpC59iEU43FagJAPLigr3AP3s38KM982bu'),
    },
    {
        symbol: 'jitoSOL',
        marketIndex: 6,
        oracle: new web3_js_1.PublicKey('7yyaeuJ1GGtVBLT2z2xub5ZWYKaNhF28mj1RdV4VDFVk'),
        oracleSource: __1.OracleSource.PYTH,
        mint: new web3_js_1.PublicKey('J1toso1uCk3RLmjorhTtrVwY9HJ7X8V9yYac6Y7kGCPn'),
        precision: new __1.BN(10).pow(numericConstants_1.NINE),
        precisionExp: numericConstants_1.NINE,
        serumMarket: new web3_js_1.PublicKey('DkbVbMhFxswS32xnn1K2UY4aoBugXooBTxdzkWWDWRkH'),
        phoenixMarket: new web3_js_1.PublicKey('5LQLfGtqcC5rm2WuGxJf4tjqYmDjsQAbKo2AMLQ8KB7p'),
    },
    {
        symbol: 'PYTH',
        marketIndex: 7,
        oracle: new web3_js_1.PublicKey('nrYkQQQur7z8rYTST3G9GqATviK5SxTDkrqd21MW6Ue'),
        oracleSource: __1.OracleSource.PYTH,
        mint: new web3_js_1.PublicKey('HZ1JovNiVvGrGNiiYvEozEVgZ58xaU3RKwX8eACQBCt3'),
        precision: new __1.BN(10).pow(numericConstants_1.SIX),
        precisionExp: numericConstants_1.SIX,
        serumMarket: new web3_js_1.PublicKey('4E17F3BxtNVqzVsirxguuqkpYLtFgCR6NfTpccPh82WE'),
        phoenixMarket: new web3_js_1.PublicKey('2sTMN9A1D1qeZLF95XQgJCUPiKe5DiV52jLfZGqMP46m'),
    },
    {
        symbol: 'bSOL',
        marketIndex: 8,
        oracle: new web3_js_1.PublicKey('AFrYBhb5wKQtxRS9UA9YRS4V3dwFm7SqmS6DHKq6YVgo'),
        oracleSource: __1.OracleSource.PYTH,
        mint: new web3_js_1.PublicKey('bSo13r4TkiE4KumL71LsHTPpL2euBYLFx6h9HP3piy1'),
        precision: new __1.BN(10).pow(numericConstants_1.NINE),
        precisionExp: numericConstants_1.NINE,
        serumMarket: new web3_js_1.PublicKey('ARjaHVxGCQfTvvKjLd7U7srvk6orthZSE6uqWchCczZc'),
    },
    {
        symbol: 'JTO',
        marketIndex: 9,
        oracle: new web3_js_1.PublicKey('D8UUgr8a3aR3yUeHLu7v8FWK7E8Y5sSU7qrYBXUJXBQ5'),
        oracleSource: __1.OracleSource.PYTH,
        mint: new web3_js_1.PublicKey('jtojtomepa8beP8AuQc6eXt5FriJwfFMwQx2v2f9mCL'),
        precision: new __1.BN(10).pow(numericConstants_1.NINE),
        precisionExp: numericConstants_1.NINE,
        serumMarket: new web3_js_1.PublicKey('H87FfmHABiZLRGrDsXRZtqq25YpARzaokCzL1vMYGiep'),
        phoenixMarket: new web3_js_1.PublicKey('BRLLmdtPGuuFn3BU6orYw4KHaohAEptBToi3dwRUnHQZ'),
    },
    {
        symbol: 'WIF',
        marketIndex: 10,
        oracle: new web3_js_1.PublicKey('6ABgrEZk8urs6kJ1JNdC1sspH5zKXRqxy8sg3ZG2cQps'),
        oracleSource: __1.OracleSource.PYTH,
        mint: new web3_js_1.PublicKey('EKpQGSJtjMFqKZ9KQanSqYXRcF8fBopzLHYxdM65zcjm'),
        precision: new __1.BN(10).pow(numericConstants_1.SIX),
        precisionExp: numericConstants_1.SIX,
        serumMarket: new web3_js_1.PublicKey('2BtDHBTCTUxvdur498ZEcMgimasaFrY5GzLv8wS8XgCb'),
        phoenixMarket: new web3_js_1.PublicKey('6ojSigXF7nDPyhFRgmn3V9ywhYseKF9J32ZrranMGVSX'),
    },
    {
        symbol: 'JUP',
        marketIndex: 11,
        oracle: new web3_js_1.PublicKey('g6eRCbboSwK4tSWngn773RCMexr1APQr4uA9bGZBYfo'),
        oracleSource: __1.OracleSource.PYTH,
        mint: new web3_js_1.PublicKey('JUPyiwrYJFskUPiHa7hkeR8VUtAeFoSYbKedZNsDvCN'),
        precision: new __1.BN(10).pow(numericConstants_1.SIX),
        precisionExp: numericConstants_1.SIX,
        phoenixMarket: new web3_js_1.PublicKey('2pspvjWWaf3dNgt3jsgSzFCNvMGPb7t8FrEYvLGjvcCe'),
        launchTs: 1706731200000,
    },
    {
        symbol: 'RNDR',
        marketIndex: 12,
        oracle: new web3_js_1.PublicKey('CYGfrBJB9HgLf9iZyN4aH5HvUAi2htQ4MjPxeXMf4Egn'),
        oracleSource: __1.OracleSource.PYTH,
        mint: new web3_js_1.PublicKey('rndrizKT3MK1iimdxRdWabcF7Zg7AR5T4nud4EkHBof'),
        precision: new __1.BN(10).pow(numericConstants_1.EIGHT),
        precisionExp: numericConstants_1.EIGHT,
        serumMarket: new web3_js_1.PublicKey('2m7ZLEKtxWF29727DSb5D91erpXPUY1bqhRWRC3wQX7u'),
        launchTs: 1708964021000,
    },
    {
        symbol: 'W',
        marketIndex: 13,
        oracle: new web3_js_1.PublicKey('H9j8CT1bFiWHaZUPMooEaxMRHdWdJ5T9CzFn41z96JHW'),
        oracleSource: __1.OracleSource.PYTH,
        mint: new web3_js_1.PublicKey('85VBFQZC9TZkfaptBWjvUw7YbZjy52A6mjtPGjstQAmQ'),
        precision: new __1.BN(10).pow(numericConstants_1.SIX),
        precisionExp: numericConstants_1.SIX,
        phoenixMarket: new web3_js_1.PublicKey('8dFTCTAbtGuHsdDL8WEPrTU6pXFDrU1QSjBTutw8fwZk'),
        launchTs: 1712149014000,
    },
    {
        symbol: 'TNSR',
        marketIndex: 14,
        oracle: new web3_js_1.PublicKey('7Cfyymx49ipGsgEsCA2XygAB2DUsan4C6Cyb5c8oR5st'),
        oracleSource: __1.OracleSource.SWITCHBOARD,
        mint: new web3_js_1.PublicKey('TNSRxcUxoT9xBG3de7PiJyTDYu7kskLqcpddxnEJAS6'),
        precision: new __1.BN(10).pow(numericConstants_1.NINE),
        precisionExp: numericConstants_1.NINE,
        phoenixMarket: new web3_js_1.PublicKey('AbJCZ9TAJiby5AY3cHcXS2gUdENC6mtsm6m7XpC2ZMvE'),
        launchTs: 1712593532000,
    },
];
exports.SpotMarkets = {
    devnet: exports.DevnetSpotMarkets,
    'mainnet-beta': exports.MainnetSpotMarkets,
};
